graphqlapi/src/main/java/io/stargate/graphql/persistence/schemafirst/SchemaSourceDao.java

```diff
@@ -0,0 +1,319 @@
+      throws Exception {
+    ensureTableExists();
+    List<BuiltCondition> conditions =
+        force
+            ? ImmutableList.of(
+                BuiltCondition.of(LATEST_VERSION_COLUMN_NAME, Predicate.EQ, expectedLatestVersion))
+            : ImmutableList.of(
+                BuiltCondition.of(LATEST_VERSION_COLUMN_NAME, Predicate.EQ, expectedLatestVersion),
+                BuiltCondition.of(DEPLOYMENT_IN_PROGRESS_COLUMN_NAME, Predicate.NEQ, true));
+    BoundQuery updateDeploymentToInProgress =
+        dataStore
+            .queryBuilder()
+            .update(KEYSPACE_NAME, TABLE_NAME)
+            .value(DEPLOYMENT_IN_PROGRESS_COLUMN_NAME, true)
+            .where(NAMESPACE_COLUMN_NAME, Predicate.EQ, namespace)
+            .ifs(conditions)
+            .build()
+            .bind();
+
+    ResultSet resultSet = dataStore.execute(updateDeploymentToInProgress).get();
+    Row row = resultSet.one();
+    if (!row.getBoolean(APPLIED_COLUMN_NAME)) {
+      boolean hasVersion =
+          row.columns().stream().anyMatch(c -> LATEST_VERSION_COLUMN_NAME.equals(c.name()));
+      if (!hasVersion) {
+        throw new IllegalStateException(
+            "You specified expectedVersion but no previous version was found");
+      }
+      UUID actualLatestVersion = row.getUuid(LATEST_VERSION_COLUMN_NAME);
+      if (Objects.equals(actualLatestVersion, expectedLatestVersion)) {
+        assert row.getBoolean(DEPLOYMENT_IN_PROGRESS_COLUMN_NAME);
+        throw new IllegalStateException(
+            "It looks like someone else is deploying a new schema, please check the latest version and try again. "
+                + "This can also happen if a previous deployment failed unexpectedly, in that case you can use the "
+                + "'force' argument to bypass this check.");
+      }
+      throw new IllegalStateException(
+          String.format(
+              "You specified expectedVersion %s, but there is a more recent version %s",
+              expectedLatestVersion, actualLatestVersion));
+    }
+  }
+
+  public void abortDeployment(String namespace) throws ExecutionException, InterruptedException {
+    BoundQuery updateDeploymentToNotInProgress =
+        dataStore
+            .queryBuilder()
+            .update(KEYSPACE_NAME, TABLE_NAME)
+            .value(DEPLOYMENT_IN_PROGRESS_COLUMN_NAME, false)
+            .where(NAMESPACE_COLUMN_NAME, Predicate.EQ, namespace)
+            .build()
+            .bind();
+    dataStore.execute(updateDeploymentToNotInProgress).get();
+  }
+
+  public void purgeOldVersions(String namespace) throws Exception {
+    List<SchemaSource> allSchemasForNamespace = getAllVersions(namespace);
+
+    int numberOfEntriesToRemove =
+        allSchemasForNamespace.size() - NUMBER_OF_RETAINED_SCHEMA_VERSIONS;
+    if (numberOfEntriesToRemove > 0) {
+      LOGGER.info("Removing {} old schema entries.", numberOfEntriesToRemove);
```

Comment:
It's maybe a bit deceptive because, since we purge on each insert, `numberOfEntriesToRemove` will always be 1 in practice. The only way to observe something higher would be to insert entries manually.

> which btw is not even the last one

Are you sure? The table is clustered by `version DESC`, and `version` is a version-1 UUID, so I think we're good. But if you can think of a failing scenario we can add a test in `SchemaDeploymentTest`.

Reply:
I saw later that you are purging on every insert.. But still the method does not do what's supposed imo.. Imagine there is exception between insert and purge, you will have extra schema hanging forever as soon as this was insert that created the 11th version.